torch/csrc/autograd/variable.h

```diff
@@ -324,6 +338,45 @@ struct TORCH_API AutogradMeta : public c10::AutogradMetaInterface {
 ///
 /// Relevant logic for both differentiable and non-differentiable views is implemented in
 /// make_variable_(non_)differentiable_view below, and wrap_output of gen_variable_type.py.
+
+
+/// NOTE [ View + Inplace detection ]
+///
+/// We want to detect views followed by inplace as they are often forbidden to ensure
+/// correctness of the computed gradients. But since we want to only notify the user
+/// when both happen, we tag the DifferentiableViewMeta when the view is created
+/// via the `make_vairable_*_view()` functions. This tag is then checked by the
+/// `check_inplace()` function from `VariableTypeUtils.h` that should be called before
+/// every inplace operation. To detect cases where other views are modified and this
+/// one is rebased by side effect, we also check in the `VariableHooks::grad_fn()`.
+
+
+/// Here are all the possible case for a DifferentiableViewMeta and how they are created.
+/// +----------------------+------------------------------------------------+------------------------------------------------+
+/// |                      | grad_fn_ == nullptr                            | grad_fn_ != nullptr                            |
+/// +----------------------+------------------------------------------------+------------------------------------------------+
+/// | WARN_REBASE          | view created by single-output in no_grad mode  | should not happen                              |
+/// |                      | (outside of custom Functions)                  |                                                |
+/// +----------------------+------------------------------------------------+------------------------------------------------+
+/// | WARN_REBASE_FUNCTION | view created during forward of custom Function | view created during forward of custom Function |
+/// |                      | in no_grad mode (including multi-output ones)  | (including multi-output ones)                  |
+/// +----------------------+------------------------------------------------+------------------------------------------------+
+/// | ERROR_REBASE         | view created by multi-output codegen or custom | view created by multi-output codegen or custom |
+/// |                      | function in no_grad mode                       | function                                       |
+/// +----------------------+------------------------------------------------+------------------------------------------------+
+/// | ALLOW_REBASE         | view created in no_grad mode from a base that  | All other differentiable views                 |
+/// |                      | do not require gradients                       |                                                |
+/// +----------------------+------------------------------------------------+------------------------------------------------+
```

Comment:
I'm not... sure what I'm supposed to do with this table. When would I be in a situation where I have a view meta with `WARN_REBASE` and I want to know how it was created? That seems like an unlikely question for me to ask. I'm probably more interested in the semantic reasoning behind all of this. Basically, suppose a disk crashed wiped out all of the source code, and I had to reconstruct this table from scratch. How would I reason my way there?

Reply:
This is the justification for the logic in `handle_view_on_rebase` in an easy to read form.
I added it here to make sure people will update it if they change this part of the code.

The semantic reasoning that lead to this is captured just above in the `Differentiable Views` note and in particular what happens when we return multiple views or what happens when the views are computed in no grad mode.

Reply:
Yeah, see my comment below. It is incredibly unlikely this is going to get updated.